### PR TITLE
Ramdisk

### PR DESCRIPTION
**Pull request by redhog**
Support an arbitrarily sized ramdisk using PMEM kernel driver.

Why not initrd? Because initrd and initramdisk have size limitations, for larger disks, it gets stuck loading the disk (probably copying and extracting the entire disk at boot). PMEM containing a normal filesystem, such as ext3 or squashfs can store an arbitrary number of files with no performance penalty at boot or when just accessing a small number of them.

### CODE REVIEW

**Comment by franzflasch** [comment]
Hi! Thanks for your PR! I really appreciate this! 

Just some remarks:
- please don't add device tree binaries to the repository. It should only contain the sources not the compilations
- Can you reorganize the device-trees please? I would like to keep the original device-trees as they are. 
- You can add a new one e.g. "riscv_em_rd.dts"
- How can such a ramdisk be built?
- If this should be added I would like to have something like a "how-to build a ramdisk for riscv_em". Otherwise this might end up being there but no one knows how to actually use it.

Thanks!

**Comment by redhog** [comment]
There are many ways to build such a ramdisk, as it's just a filesystem image. For squashfs, you

```
    cd filesystemtree; mksquashfs ./ ../myimage.squashfs -Xcompression-level 9
```

For ext3 etc, you'd have to make an empty file of the right size, then run `mkfs.ext3 myimage.ext3`, then `mount -o loop myimage.ext3 /mnt; cp -a filesystemtree/* /mnt; umount /mnt`.

So pretty straight forward to make (but what files to put in it is entirely up to the user, if you want to use it as a root disk, obviously you'd need to use buildroot or something).

**Comment by redhog** [comment]
Added some docs on the ramdisk usage, including explaining why this uses pmem over initrd/initramdisk.

File: src/core/riscv_config.h [hunk 1]
```diff
@@ -9,8 +9,10 @@
 #define MROM_BASE_ADDR 0x1000UL
 #define MROM_SIZE_BYTES 0xf000UL
 
-#define RAM_BASE_ADDR 0x80000000UL
-#define RAM_SIZE_BYTES 0x8000000UL /* 128MB such as the default for the qemu virt machine */
+// #define RAM_BASE_ADDR 0x80000000UL
+// #define RAM_SIZE_BYTES 0x8000000UL /* 128MB such as the default for the qemu virt machine */
+#define RAM_BASE_ADDR          0x80000000UL
+#define RAM_SIZE_BYTES         0x40000000UL
```
Comment (franzflasch) [inline src/core/riscv_config.h:15]: Why the increase of the main memory? Please keep the old setting.

File: src/main.c [hunk 2]
```diff
@@ -113,6 +115,15 @@ static void parse_options(int argc,
                 // }
                 break;
             }
+            case 'i':
+            {
+                arg_initrd_file = optarg;
+                // if (arg_fw_file)
```
Comment (franzflasch) [inline src/main.c]: Please remove the commented code here. I know I am doing that on my own on a lot of places, but we should not introduce new ones.

File: src/soc/riscv_example_soc.c [hunk 3]
```diff
@@ -113,6 +116,11 @@ void rv_soc_init(rv_soc_td *rv_soc, char *fw_file_name, char *dtb_file_name)
 
     write_mem_from_file(fw_file_name, soc_ram, sizeof(soc_ram));
 
+    if (initrd_file_name != NULL) {
+     printf("XXXXXXXXXXXXXXXXXXXXXXXXXXX\n");
```
Comment (franzflasch) [inline src/soc/riscv_example_soc.c]: Please remove this print statement or change it to something meaningful.

File: dts/riscv_em.dts [hunk 4]
```diff
@@ -31,9 +31,14 @@
 
     sram: memory@80000000 {
         device_type = "memory";
-        reg = <0x0 0x80000000 0x0 0x8000000>;
+        reg = <0x0 0x80000000 0x0 0x40000000>;
```
Comment (franzflasch) [inline dts/riscv_em.dts:38]: Why the increase of the main memory? Please keep the old setting.

File: src/core/riscv_config.h (hunk 1 above)
Comment (redhog) [reply src/core/riscv_config.h:15]: Oups, that was just left in there from other experimentation (using initrd). Removing.

File: src/soc/riscv_example_soc.c (hunk 3 above)
Comment (redhog) [reply src/soc/riscv_example_soc.c]: Oups... again, leftover from experimentation...

**Pull request merged**